src/Microsoft.OData.Client/BulkUpdateSaveResult.cs

```diff
@@ -202,19 +253,329 @@ protected override ODataRequestMessageWrapper CreateRequestMessage(string method
+            }
+            else if (materializer is MaterializerEntry entry)
+            {
+                return entry.Entry.InstanceAnnotations.Any(a => a.Name == XmlConstants.DataModificationType);
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        /// This method creates an operation response for each operation.
+        /// </summary>
+        /// <param name="materializer">An instance of the <see cref="IMaterializerState"/> entry associated with an operation response.</param>
+        /// <param name="isTopLevelResponse">true for a top-level response otherwise false.</param>
+        /// <exception cref="DataServiceRequestException"></exception>
+        private void CreateOperationResponse(IMaterializerState materializer, bool isTopLevelResponse)
+        {
+            Exception ex = null;
+            OperationResponse operationResponse = null;
+            DataServiceResponse service = new DataServiceResponse(null, -1, this.operationResponses, batchResponse: false);
+            Descriptor descriptor = null;
+
+            try
+            {        
+                if (this.lastVisitedDescriptor == null)
+                {
+                    descriptor = this.FindNonVisitedDescriptor(this.bulkUpdateGraph.TopLevelDescriptors);
+                    this.descriptorsAdjacentToVisitedDescriptor.Push(this.bulkUpdateGraph.TopLevelDescriptors);
+                }
+                else
+                {
+                    HashSet<Descriptor> relatedDescriptors = this.bulkUpdateGraph.GetRelatedDescriptors(lastVisitedDescriptor);
+                    
+                    if (relatedDescriptors.Count == 0)
+                    {
+                        if (this.descriptorsAdjacentToVisitedDescriptor != null && this.descriptorsAdjacentToVisitedDescriptor.Peek().Except(this.visited).ToList().Count == 0)
+                        {
+                            this.descriptorsAdjacentToVisitedDescriptor.Pop();
+                        }
+
+                        descriptor = this.FindNonVisitedDescriptor(this.descriptorsAdjacentToVisitedDescriptor.Peek());
+                    }
+                    else 
+                    {
+                        this.descriptorsAdjacentToVisitedDescriptor.Push(relatedDescriptors);
+                        descriptor = this.FindNonVisitedDescriptor(relatedDescriptors);
+                    }
+                }
+
+                this.lastVisitedDescriptor = descriptor;
+                 
+                if (descriptor != null) 
+                {
+                    this.SaveResultProcessed(descriptor, HasFailedOperation(materializer));
+
+                    operationResponse = new ChangeOperationResponse(this.cachedResponse.Headers, descriptor)
+                    {
+                        StatusCode = (int)this.cachedResponse.StatusCode
+                    };
+
+                    this.descriptorForMaterializer.Add(descriptor, materializer);

```

Comment:
The thinking here is that the structure of the `MaterializerDeltaFeed` will be the exact same as the structure of the `BulkUpdateGraph` that was used in creating the request. 
![Uploading Screenshot 2023-04-03 141639.png…]()

Reply:
The thinking here is that the structure of the `MaterializerDeltaFeed` will be the exact same as the structure of the `BulkUpdateGraph` that was used in creating the request. 
![Screenshot 2023-04-03 141639](https://user-images.githubusercontent.com/25525526/229494850-4374af6a-b2cd-4019-bae1-1b6a7e901da8.png)

Reply:
The thinking here is that the structure of the MaterializerDeltaFeed will be the exact same as the structure of the BulkUpdateGraph that was used in creating the request. This will mean that when we start looping through the `MaterializerFeed` object, we'll be picking the descriptor on the `BulkUpdateGraph` that corresponds to the position of the `MaterializerFeed`

Reply:
I have updated this logic and used lists and indexed the lists to ensure the orrder remains the same.